scripts/build_perf.go

```diff
@@ -0,0 +1,192 @@
+// Copyright 2017 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+// Measure the performance of building all the Go commands under various GOGC
+// values. The output is four csv files:
+//
+// - build_perf_real.csv
+// - build_perf_user.csv
+// - build_perf_sys.csv
+// - build_perf_max_rss.csv
+package main
+
+import (
+	"encoding/csv"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"sync"
+	"syscall"
+	"time"
+)
+
+const (
+	cmdsPath  = "$GOPATH/src/github.com/u-root/u-root/cmds"
+	gogcBegin = 50
+	gogcEnd   = 2000
+	gogcStep  = 50
+)
+
+// The fields profiled by a single `go build`.
+type measurement struct {
+	realTime float64 // seconds
+	userTime float64 // seconds
+	sysTime  float64 // seconds
+	maxRss   int64   // KiB
+}
+
+// Each CSV file only stores one field of the measurement.
+// This struct describes a single CSV file.
+type csvDesc struct {
+	filename string
+	field    func(measurement) string
+	// Measurements are sent from the `measureBuilds` to the `writeCsv`
+	// functions via this channel.
+	c chan []*measurement
+}
+
+var descs = []csvDesc{
+	{
+		"build_perf_real.csv",
+		func(m measurement) string { return fmt.Sprint(m.realTime) },
+		make(chan []*measurement),
+	}, {
+		"build_perf_user.csv",
+		func(m measurement) string { return fmt.Sprint(m.userTime) },
+		make(chan []*measurement),
+	}, {
+		"build_perf_sys.csv",
+		func(m measurement) string { return fmt.Sprint(m.sysTime) },
+		make(chan []*measurement),
+	}, {
+		"build_perf_max_rss.csv",
+		func(m measurement) string { return fmt.Sprint(m.maxRss) },
+		make(chan []*measurement),
+	},
+}
+
+var wg sync.WaitGroup
+
+// Return a list of command names.
+func getCmdNames() ([]string, error) {
+	files, err := ioutil.ReadDir(os.ExpandEnv(cmdsPath))
+	if err != nil {
+		return nil, err
+	}
+	cmds := []string{}
+	for _, file := range files {
+		if file.IsDir() {
+			cmds = append(cmds, file.Name())
+		}
+	}
+	return cmds, nil
+}
+
+func buildCmd(cmd string, gogc int) (*measurement, error) {
+	start := time.Now()
+	c := exec.Command("go", "build")
```

Comment:
Sounds good. I'm planning on applying the `CompileInTempDir` en masse to all the existing tests in its own PR.